xdsl/analysis/sparse_analysis.py

```diff
@@ -231,3 +232,147 @@ def __str__(self) -> str:
 
 
 PropagatingLatticeInvT = TypeVar("PropagatingLatticeInvT", bound=PropagatingLattice)
+
+
+class SparseForwardDataFlowAnalysis(DataFlowAnalysis, Generic[PropagatingLatticeInvT]):
+    """
+    Base class for sparse forward data-flow analyses. It propagates lattices
+    attached to SSA values along the direction of data flow.
+    """
+
+    def __init__(
+        self, solver: DataFlowSolver, lattice_type: type[PropagatingLatticeInvT]
+    ):
+        super().__init__(solver)
+        self.lattice_type = lattice_type
+
+    def initialize(self, op: Operation) -> None:
+        # Set entry state for all arguments of the top-level regions.
+        for region in op.regions:
+            if region.first_block is not None:
+                for arg in region.first_block.args:
+                    self.set_to_entry_state(self.get_lattice_element(arg))
+
+        # Recursively visit all ops to build initial dependencies.
+        def initialize_recursively(op: Operation):
+            self.visit(ProgramPoint.before(op))
+            for region in op.regions:
+                for block in region.blocks:
+                    block_start_point = ProgramPoint.at_start_of_block(block)
+                    executable = self.get_or_create_state(block_start_point, Executable)
+                    executable.block_content_subscribers.add(self)
+                    self.visit(block_start_point)
+                    for nested_op in block.ops:
+                        initialize_recursively(nested_op)
+
+        initialize_recursively(op)
+
+    def visit(self, point: ProgramPoint) -> None:
+        if point.op is not None:
+            self.visit_operation(point.op)
+        elif point.block is not None:
+            # This case handles end-of-block points, which for forward analysis
+            # means we are visiting the block itself to handle its arguments.
+            self.visit_block(point.block)
+
+    def visit_operation(self, op: Operation) -> None:
+        """Transfer function for an operation's results."""
+        if not op.results:
+            return
+
+        # If the parent block is not executable, do nothing.
+        if (
+            op.parent
```

Comment:
Quite important especially if we one day make Block falsey if empty
```suggestion
            op.parent is not None
```